pkg/execution/executor/executor.go

```diff
@@ -2455,8 +2455,9 @@ func (e *executor) handleGeneratorStep(ctx context.Context, i *runInstance, gen
 		Payload:               queue.PayloadEdge{Edge: nextEdge},
 		Metadata:              make(map[string]any),
 	}
+	nextItem.SetOptimizedParallelism(gen.OptimizedParallelism())
 
-	if !hasPendingSteps {
+	if !hasPendingSteps || !i.item.OptimizedParallelism() {
```

Comment:
`if !hasPendingSteps` felt bad to have everywhere, especially seeing as it's such a common check.

Would it be reasonable to have a helper like:
```go
func shouldEnqueueDiscovery(hasPendingSteps bool, qi queue.Item) bool {
  return !hasPendingSteps || !nextItem.item.OptimizedParallelism()
}

// ...

if shouldEnqueueDiscovery(hasPendingSteps, qi) {
  // ...
}
```

It's a total nit and feels a bit over the top, but at least then if we require more things to check etc we're not having to trawl through the code to find where we `Enqueue()`.

Maybe there are a few spots where it just doesn't make sense too. 'tis a nit - feel free to ignore.